File: Others/Python3/268.Missing Number.py
"""
@Title: 268. Missing Number
@Tag: array
@Date: Jan-01 2020
@Author: ceezyyy
@Difficulty: Easy
"""


# Approach 1: (Sum)


class Solution:
    def missingNumber(self, nums: List[int]) -> int:
        return sum(range(len(nums)+1)) - sum(nums)


"""
Runtime: 136 ms, faster than 91.51% of Python3 online submissions for Missing Number.
Memory Usage: 14 MB, less than 90.32% of Python3 online submissions for Missing Number.
"""


"""
Time Complexity: O(n)
Space Complexity: O(1)
"""


# Approach 2: (Sort)


class Solution:
    def missingNumber(self, A: List[int]) -> int:
        A = sorted(A)
        if A[0] != 0:  # corner case
            return 0
        n = len(A)
        for i in range(1, n):
            if A[i] != A[i-1] + 1:
                return A[i] - 1
        return A[-1] + 1


"""
Runtime: 156 ms, faster than 24.68% of Python3 online submissions for Missing Number.
Memory Usage: 14 MB, less than 83.87% of Python3 online submissions for Missing Number.
"""


"""
Time Complexity: O(nlogn)
Space Complexity: O(1)
"""

File: others/cpp/268.Missing Number.cpp
/**
 * Source: LeetCode 268. Missing Number.cpp
 * Author: ceezyyy
 * Date: 03-22-2019
 */


/*
Easy

Share
Given an array containing n distinct numbers taken from 0, 1, 2, ¡­, n, 
find the one that is missing from the array.

Example 1:
Input: [3,0,1]
Output: 2

Example 2:
Input: [9,6,4,2,3,5,7,0,1]
Output: 8

Note:
Your algorithm should run in linear runtime complexity. Could you implement it using only constant extra space complexity?

*/


#include<iostream>
#include<vector>
#include<map>
using namespace std;

class Solution {
public:
	int missingNumber(vector<int>& nums) {
		map<int, int> count;
		for (int i = 0; i < nums.size(); i++) {
			count[nums.at(i)]++;
		}
		for (int j = 0; j <= nums.size(); j++) {
			/*
			Here is 'j<=nums.size()'
			because the array containing n distinct numbers
			taken from 0, 1, 2, ..., n
			*/
			if (count[j] == 0) {
				return j;
			}
		}
		return 0;
	}
};


/**
 * Submission Detail:
 * Runtime: 68 ms, faster than 6.36% of C++ online submissions for Missing Number.
 * Memory Usage: 17.5 MB, less than 5.19% of C++ online submissions for Missing Number.
 */
